static/js/passed-state.js

```diff
@@ -0,0 +1,115 @@
+export default class PassedState {
+    _key = 'python-type-challenges';
+
+    /**
+     * Initializing when there is no state in the local storage. If there is no state in the local storage, the initial state is required.
+     * this function will check the new state and the old state whether is undefined or not and updated the old state based on new state.
+     * 
+     * @param {object} newState - the initial state of the challenges which grouped by the level.
+     * @returns void
+     */
+    init(newState) {
+        const oldState = this.get();
+        // initialize the state when there is no state in the local storage.
+        if (!oldState && !newState) {
+            throw new Error('initial state is required when there is no state in the local storage.');
+        }
+
+        // check new state and old state whether is undefined or not. and merge the new state to the old state.
+        const state = this._checkAndMerge(oldState, newState);
+        this._save(state);
+    }
+
+    get() {
+        const currentState = localStorage.getItem(this._key);
+        return JSON.parse(currentState);
+    }
+
+    /**
+     * Save the state to the local storage with JSON format.
+     * @param {object} state - the state contains the challenge name and whether the challenge is passed.
+     */
+    _save(state) {
+        localStorage.setItem(this._key, JSON.stringify(state));
+    }
+
+    /**
+     * Set the target challenge as passed in the state.
+     * 
+     * @param {'basic' | 'intermediate' | 'advanced' | 'extreme'} level - the level of the challenge.
+     * @param {string} challengeName - the name of the challenge.
+     * @returns void
+     */
+    setPassed(level, challengeName) {
+        let state = this.get();
+
+        const challenges = state[level];
```

Comment:
How about storing a `Map` as the state, this way we don't need to iterate overall every challenge just to mutate one item